src/core/connection_pool.c

```diff
@@ -0,0 +1,521 @@
+    if (ServerNameLength == QUIC_MAX_SNI_LENGTH + 1) {
+        Status = QUIC_STATUS_INVALID_PARAMETER;
+        QuicTraceLogError(
+            ConnPoolServerNameTooLong,
+            "[conp] ServerName too long (%llu bytes)",
+            ServerNameLength);
+        return Status;
+    }
+
+    QuicTraceEvent(
+        ApiEnter,
+        "[ api] Enter %u (%p).",
+        QUIC_TRACE_API_CONNECTION_POOL_CREATE,
+        Config->Registration);
+
+    CxPlatZeroMemory(ConnectionPool, sizeof(HQUIC) * Config->NumberOfConnections);
+
+    //
+    // Resolve the server name or use the remote address.
+    //
+    QUIC_ADDR ResolvedRemoteAddress = { 0 };
+    QuicAddrSetFamily(&ResolvedRemoteAddress, Config->Family);
+    if (Config->ServerAddress != NULL) {
+         ResolvedRemoteAddress = *Config->ServerAddress;
+    } else {
+        Status =
+            CxPlatDataPathResolveAddress(
+                MsQuicLib.Datapath,
+                Config->ServerName,
+                &ResolvedRemoteAddress);
+        if (QUIC_FAILED(Status)) {
+            QuicTraceLogError(
+                ConnPoolResolveAddress,
+                "[conp] Failed to resolve address, 0x%x",
+                Status);
+            goto Error;
+        }
+    }
+
+    QuicAddrSetPort(&ResolvedRemoteAddress, Config->ServerPort);
+
+    //
+    // Get the local address and a port to start from.
+    //
+    QUIC_ADDR LocalAddress;
+    Status = QuicConnPoolGetStartingLocalAddress(&ResolvedRemoteAddress, &LocalAddress);
+    if (QUIC_FAILED(Status)) {
+        QuicTraceLogError(
+            ConnPoolGetLocalAddress,
+            "[conp] Failed to get local address, 0x%x",
+            Status);
+        goto Error;
+    }
+
+    uint32_t InterfaceIndex;
+    Status = QuicConnPoolGetInterfaceIndexForLocalAddress(&LocalAddress, &InterfaceIndex);
+    if (QUIC_FAILED(Status)) {
+        goto Error;
+    }
+
+    Status = CxPlatDataPathRssConfigGet(InterfaceIndex, &RssConfig);
+    if (QUIC_FAILED(Status)) {
+        QuicTraceLogError(
+            ConnPoolGetRssConfig,
+            "[conp] Failed to get RSS config, 0x%x",
+            Status);
+        goto Error;
+    }
+
+    if (RssConfig->RssIndirectionTableLength < sizeof(uint32_t)) {
+        //
+        // No RSS cores configured.
+        //
+        Status = QUIC_STATUS_INVALID_STATE;
+        QuicTraceLogError(
+            ConnPoolRssNotConfigured,
+            "[conp] RSS not configured, 0x%x",
+            Status);
+        goto Error;
+    }
+
+    if (RssConfig->RssSecretKeyLength > CXPLAT_TOEPLITZ_KEY_SIZE_MAX) {
+        Status = QUIC_STATUS_INTERNAL_ERROR;
+        QuicTraceLogError(
+            ConnPoolRssSecretKeyTooLong,
+            "[conp] RSS secret key too long (%u bytes), 0x%x",
+            RssConfig->RssSecretKeyLength,
+            Status);
+        goto Error;
+    } else if (RssConfig->RssSecretKeyLength < CXPLAT_TOEPLITZ_KEY_SIZE_MIN) {
+        Status = QUIC_STATUS_INVALID_STATE;
```

Comment:
Shouldn't this one also be `INTERNAL_ERROR`?
Is there a scenario that can lead here, or is this check to be cautious but essentially an assert?

Reply:
I see...
`INVALID_STATE` is generally used when the caller did something at the wrong time or in the wrong order.

Do you know if the user could do something that affect the rss key length? If yes, INVALID_STATE can make sense.
But if it means the OS is in a wrong config (or a config ew don't expect), I think I'd keep INTERNAL_ERROR (or some other more precise error code)